### PR TITLE
Changing something to the domain should leave the domain_version untouched

### PR DESCRIPTION
**Pull request by johanv**
This is a unit  test and a fix for CRM-17430.

---
- [CRM-17430: Altering the domain using the API changes the CiviCRM version to 3](https://issues.civicrm.org/jira/browse/CRM-17430)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
I agree with this change in principle - but looks like there is something else that needs to be done to get the tests to pass.

**Comment by johanv** [comment]
Hmm, I guess it does make sense for domain_version to be required. Maybe I should revert my last commit and change some tests.

**Comment by johanv** [comment]
I think the failing test of the latest commit is not caused by the changes in that commit.

**Comment by eileenmcnaughton** [comment]
Am I right in thinking version is no longer a mandatory field (ie. the fields that are mandatory in the spec function are mandatory for create not update).

Yes, failure is unrelated

**Comment by johanv** [comment]
When you create a new domain, and you don't provide a version, would the api use a sensible default version? If so, the version doesn't have to be mandatory for create, and the PR is good as it is.

**Comment by eileenmcnaughton** [comment]
I agree that a sensible default version makes sense - but I don't *think* that is already happening? Or am I not up-to-date on this?

**Comment by colemanw** [comment]
I feel like I'm missing something... why is the fix not simply

    function _civicrm_api3_domain_create_spec(&$params) {
      $params['version']['api.aliases'] = array('domain_version');
    }

**Comment by johanv** [comment]
That might work, but I'm not sure what will happen, because civicrm_api3 will add 'version' => 3 to the params. I'm sorry that I can't try this out right now, the weekend is more busy than I expected.

**Comment by seamuslee001** [comment]
Jenkins re test this please

**Comment by johanv** [comment]
I put

     $params['domain_version']['api.required'] = 1;

back in `_civicrm_api3_domain_create_spec`, because it was there before I started changing things. But then I had to change some unit tests, so that they provide a domain_version when creating a domain.

Now if you don't want domain_version to be mandatory, you can also merge commit 6a22976.

File: api/v3/Domain.php [hunk 1]
```diff
@@ -129,8 +133,19 @@ function _civicrm_api3_domain_get_spec(&$params) {
...
+  else {
+    unset($params['version']);
+  }
+  $result = _civicrm_api3_basic_create(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+
+  // Rename version to domain_version, see CRM-17430.
+  $result_value = CRM_Utils_Array::first($result['values']);
+  $result_value['domain_version'] = $result_value['version'];
+  unset($result_value['version']);
+  return $result;
```
Comment (colemanw) [inline api/v3/Domain.php]: I don't see how the above four lines of code could possibly work. CRM_Utils_Array::first does not return a reference.

File: api/v3/Domain.php [hunk 2]
```diff
@@ -142,7 +157,10 @@ function civicrm_api3_domain_create($params) {
  *   Array of parameters determined by getfields.
  */
 function _civicrm_api3_domain_create_spec(&$params) {
-  $params['domain_version'] = $params['version'];
+  $params['domain_version'] = array(
+    'title' => "CiviCRM Version",
+    'description' => "The civicrm version this instance is running",
+  );
```
Comment (colemanw) [inline api/v3/Domain.php]: also need to set type = string

File: tests/phpunit/api/v3/DomainTest.php [hunk 3]
```diff
@@ -176,6 +176,29 @@ public function testCreate() {
...
+      'id' => $domain_result['id'],
+      'name' => 'B-Team domain',
+    ));
+
+    // Get domain again.
+    $domain_after = $this->callAPISuccess('Domain', 'getsingle', array('id' => $domain_result['id']));
+
+    // Version should still be the same.
+    $this->assertEquals($domain_before['version'], $domain_after['version']);
+  }
```
Comment (colemanw) [inline tests/phpunit/api/v3/DomainTest.php]: @eileenmcnaughton shouldn't the syntaxConformanceTests be doing this?

Comment (eileenmcnaughton) [inline tests/phpunit/api/v3/DomainTest.php]: if it's enabled for the entity probably - but version is such an odd field I'm not sure it makes sense to rely on syntax conformance for it

**Comment by colemanw** [comment]
@johanv can you please update this PR based on the above couple of suggestions?

**Comment by johanv** [comment]
Sorry, I had health problems for 2 weeks. And now I have some catching up to do for my work. Maybe after that.

**Comment by eileenmcnaughton** [comment]
I hope you are better now!

**Comment by colemanw** [comment]
@johanv are you able to revisit this?

**Comment by johanv** [comment]
I will look at this next week.

**Comment by johanv** [comment]
I left the testUpdateDomainName test as it was, because syntaxConformanceTests did not catch the issue. (And the workings of syntaxConformanceTests are unclear to me ;))

**Comment by colemanw** [comment]
jenkins, retest this please

**Comment by colemanw** [comment]
jenkins, retest this please

**Comment by colemanw** [comment]
jenkins, retest this please

**Comment by colemanw** [comment]
jenkins, retest this please

**Comment by JoeMurray** [comment]
@twomice would you be interested in reviewing this PR?

**Comment by twomice** [comment]
Yes, thanks for thinking of me. Will do today.

**Comment by twomice** [comment]
Jenkins, retest this please.

**Comment by twomice** [comment]
This looks good to me. I merged the changes into master on a local dev environment, and tried modifying  domain records with `drush civicrm-api domain.create`. Before the merge, there's no way to modify the domain without the domain version being changed to '3' (thus reproducing the bug). After the merge, changes work as expected without changing the domain version. 

Also ran tests in api_v3_DomainTest and api_v3_SettingTest (including new and modified tests), all passing.

@JoeMurray Do I ask you to "please merge" or what's the protocol for marking this as reviewed?

**Comment by JoeMurray** [comment]
Could you also review the code against the Standard Types of Concerns on the Reviewer's Tips tab of https://docs.google.com/spreadsheets/d/1fmHFOZ83ectSPCMWvXwCeJgrw4KpYi8JEV2ZDkd6XDU/edit#gid=616660117 and check for any unanswered questions from thread above? If everything looks good then yes, ask @yashodha to merge. Thanks so much! Let me know if you could do another.

**Comment by eileenmcnaughton** [comment]
Erm & everyone needs to try to remember to close the JIRA ticket & update the fix version when merged (I forget so often that I assume others must too)

**Comment by twomice** [comment]
Thanks, @JoeMurray I'll give it another look then.

File: api/v3/Domain.php [hunk 4]
```diff
@@ -129,8 +133,22 @@ function _civicrm_api3_domain_get_spec(&$params) {
  * @return array
  */
 function civicrm_api3_domain_create($params) {
-  $params['version'] = $params['domain_version'];
-  return _civicrm_api3_basic_create(_civicrm_api3_get_BAO(__FUNCTION__), $params);
+  if (isset($params['domain_version'])) {
```
Comment (twomice) [inline api/v3/Domain.php]: @johanv This if-else block is good, and it also adds code complexity which should be matched with a test. Can you add a test that changes the domain_version via api?

**Comment by eileenmcnaughton** [comment]
I think I'm going to merge this despite the comment about another test - I think this line of the test in the PR covers the version not being incorrectly updated & this is mostly about preventing that. It would be good to test that it can be correctly updated too I guess but that is very edge case in real usage & @twomice did test manually & cover off the functionality & safety aspects of the QA..


https://github.com/civicrm/civicrm-core/pull/7047/files#diff-55bdb1297e78c86da55c1c37e3d5d959R198

**Comment by twomice** [comment]
Thanks @eileenmcnaughton. This is good and useful code and if Eileen says it's OK without another test, I'm all in. :-)

**Comment by eileenmcnaughton** [comment]
Yeah - I LIKE tests - but I don't want to make the bar too high since you & Johan have both put work into this & it DOES add tests and it has been hanging around for a long time
